File: com/white/WordBreak.java
package com.white;

import java.util.HashSet;
import java.util.List;

public class WordBreak {

    // 139. Word Break
    public boolean wordBreak(String s, List<String> wordDict) {
        
        HashSet<String> hs = new HashSet<>();
        for (String w : wordDict) {
            hs.add(w);
        }

        boolean[] flags = new boolean[s.length()+1];
        flags[0] = true;
        for (int i = 1; i <= s.length(); i++) {
            for(int j=0; j<i; j++) {
                if (flags[j] && hs.contains(s.substring(j,i))) {
                    flags[i] = true;
                    break;
                }
            }    
        }

        return flags[s.length()];
    }
} 

File: LeetcodeCore/WordBreak.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetcodeCore
{
    public class WordBreakSolution
    {
        // 139. Word Break
        public bool WordBreak(string s, IList<string> wordDict)
        {
            var hs = new HashSet<string>();
            foreach (string w in wordDict)
            {
                hs.Add(w);
            }

            var flags = new bool[s.Length + 1];
            flags[0] = true;
            for (int i = 1; i <= s.Length; i++)
            {
                for (int j = 0; j < i; j++)
                {
                    if (flags[j] && hs.Contains(s.Substring(j, i - j)))  // s[j..i] is equivalent of s.Substring(j, i - j)
                    {
                        flags[i] = true;
                        break;
                    }
                }
            }

            return flags[s.Length];
        }
    }
}
